### DISCUSSION TITLE
Trouble overfitting a small batch of block detection in webpages

### DISCUSSION THREAD

**Discussion by salman1993** [ultralytics/yolov5#2821] [General]
I am trying to use YoloV5 to recognize section blocks in webpages. For starters, I am trying to overfit to a small batch - here is a sample:
![test_batch0_labels](https://user-images.githubusercontent.com/7154255/115060960-7dfe8200-9eb6-11eb-91dd-4e44a61dfd63.jpg)

I followed the [tutorial here](https://github.com/ultralytics/yolov5/wiki/Train-Custom-Data). I was able to train the model but it is not able to predict blocks confidently. At conf_thresh 0.25, there are no block predictions but if I set it to really low value such as 0.03, I see mostly mispredictions. 

What I tried so far (nothing worked):
- playing around with data augmentation and also turning off all data augmentation -> since we can assume webpages to be upright, i thought this might be a good idea
- hyperparam evolution
- adding more training samples (started with batch of 5, tried with 60)
- training with pretrained weights and from scratch
- training with the larger models and also larger img_size (600, 1280, 1600)
- rectangular training
- different epochs between 3 to 300

Any pointers to what I can try next with YoloV5 to solve this task will be very helpful. I will put a lot information in the thread if anyone wants to take a deeper look. Thanks for making this repo so accessible!

**Comment by salman1993** [comment]
Here is the small dataset I have used:
[articles_dataset.zip](https://github.com/ultralytics/yolov5/files/6326942/articles_dataset.zip)

Btw, I also tried overfitting using some people images from Google and it worked really well (my guess is that transfer learning from COCO works a lot better on people than webpage blocks) so my suspicion was that the anchors might be an issue since the rectangular block bboxes can be quite wide and long. However, the output from autoanchor tells me that best possible recall is 1.00 after running k-means so this is fine? (not sure):
```
autoanchor: Analyzing anchors... anchors/target = 0.00, Best Possible Recall (BPR) = 0.0000. Attempting to improve anchors, please wait...
autoanchor: Running kmeans for 9 anchors on 250 points...
autoanchor: thr=0.25: 0.9960 best possible recall, 6.04 anchors past thr
autoanchor: n=9, img_size=1280, metric_all=0.378/0.755-mean/best, past_thr=0.488-mean: 191,76,  329,78,  485,159,  790,145,  667,334,  1185,233,  522,639,  1204,523,  1171,870
autoanchor: Evolving anchors with Genetic Algorithm: fitness = 0.7782: 100%|
autoanchor: thr=0.25: 1.0000 best possible recall, 6.13 anchors past thr
autoanchor: n=9, img_size=1280, metric_all=0.382/0.778-mean/best, past_thr=0.487-mean: 223,50,  317,86,  745,96,  443,176,  673,263,  1102,208,  491,639,  1123,421,  1213,768
autoanchor: New anchors saved to model. Update model *.yaml to use these anchors in the future
```

My dataset yaml file:
```
train: ../articles_dataset/images
val: ../articles_dataset/images

# number of classes
nc: 1

# class names
names: ['block']
```

My hyperparams file after running hyperparam evolution:
```
# Hyperparameter Evolution Results
# Generations: 349
#                   P         R     mAP.5 mAP.5:.95       box       obj       cls
# Metrics:      0.158      0.25     0.171    0.0397    0.0717     0.112         0

lr0: 0.0114
lrf: 0.236
momentum: 0.976
weight_decay: 0.00048
warmup_epochs: 3.23
warmup_momentum: 0.917
warmup_bias_lr: 0.0751
box: 0.0374
cls: 0.451
cls_pw: 1.01
obj: 0.953
obj_pw: 1.02
iou_t: 0.2
anchor_t: 4.7
anchors: 2.0
fl_gamma: 0.0
hsv_h: 0.0137
hsv_s: 0.9
hsv_v: 0.324
degrees: 0.0
translate: 0.136
scale: 0.363
shear: 0.0
perspective: 0.0
flipud: 0.0
fliplr: 0.5
mosaic: 0.971
mixup: 0.0
```

F1 curve of single training run (although most look quite similar):
![F1_curve](https://user-images.githubusercontent.com/7154255/115062611-8952ad00-9eb8-11eb-8850-2738e77c617c.png)

**Reply by glenn-jocher** [reply]
@salman1993 training with a single batch is not recommended. You need thousands of iterations to produce any results, therefore train with 5k, 10k etc epochs if you expect results with a single batch. Below are our full training recommendations.

👋 Hello! Thanks for asking about improving training results. Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users first train with all default settings before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥1.5k images per class
- **Instances per class.** ≥10k instances (labeled objects) per class total
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total).

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x will produce better results in nearly all cases, but have more parameters and are slower to run. For mobile applications we recommend YOLOv5s/m, for cloud or desktop applications we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

To start training from pretrained weights simply pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
```

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://user-images.githubusercontent.com/26833433/103595982-ab986000-4eb1-11eb-8c57-4726261b0a88.png"></p>

## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyp.scratch.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://github.com/ultralytics/yolov5/issues/607).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains:
http://karpathy.github.io/2019/04/25/recipe/
